### PR TITLE
Drop support for EOL Python 2

### LINKED ISSUE #725: [HELP WANTED] Clean up python 2 coding
## 🚀 Feature

Since we are dropping the python2 support, we should clean up python 2 coding style. For example, the dependency on six should be removed. PRs from OSS are welcome and will be merged.

### PR DESCRIPTION
**Pull request by hugovk**
Fixes #725.

Removes six library and upgrades Python syntax.

Also adds `python_requires` to `setup.py` to help pip, and Trove classifiers to show support on PyPI.

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
Thanks for the contribution.

File: examples/text_classification/iterable_train.py [hunk 1]
```diff
@@ -148,7 +147,7 @@ def get_csv_iterator(data_path, ngrams, vocab, start=0, num_lines=None):
     """
     def iterator(start, num_lines):
         tokenizer = get_tokenizer("basic_english")
-        with io.open(data_path, encoding="utf8") as f:
+        with open(data_path, encoding="utf8") as f:
```
Comment (zhangguanheng66) [inline examples/text_classification/iterable_train.py:150]: Is `io.open` function, rather than built-in `open` function, recommended for Python3?

File: examples/text_classification/iterable_train.py [hunk 2]
```diff
@@ -176,7 +175,7 @@ class Dataset(torch.utils.data.IterableDataset):
         num_lines: the number of lines read by the individual iterator.
     """
     def __init__(self, iterator, num_lines):
-        super(Dataset, self).__init__()
+        super().__init__()
```
Comment (zhangguanheng66) [inline examples/text_classification/iterable_train.py:178]: Is this fix relevant to the drop of python 2 support? If not, I suggest to fix in a separate issue.

File: examples/text_classification/iterable_train.py [hunk 3]
```diff
@@ -210,15 +209,14 @@ def __iter__(self):
         if self._setup is False:
             self._setup_iterator()
             self._setup = True
-        for x in self._iterator:
-            yield x
+        yield from self._iterator
```
Comment (zhangguanheng66) [inline examples/text_classification/iterable_train.py:212]: Is this fix relevant to the drop of python 2 support? If not, I suggest to fix in a separate issue.

File: examples/text_classification/iterable_train.py [hunk 4]
```diff
@@ -210,15 +209,14 @@ def __iter__(self):
...
 
 
 def count(data_path):
     r"""
     return the total numerber of text entries and labels.
     """
-    with io.open(data_path, encoding="utf8") as f:
```
Comment (zhangguanheng66) [inline examples/text_classification/iterable_train.py:221]: Same as above

File: examples/text_classification/iterable_train.py [hunk 5]
```diff
@@ -210,15 +209,14 @@ def __iter__(self):
...
+    with open(data_path, encoding="utf8") as f:
```
Comment (zhangguanheng66) [inline examples/text_classification/iterable_train.py:219]: see above

File: test/test_vocab.py [hunk 6]
```diff
@@ -101,7 +99,7 @@ def test_vocab_download_fasttext_vectors(self):
         # to test string aliases.
         for i in range(3):
             if i == 2:
-                vectors = str("fasttext.simple.300d")  # must handle str on Py2
+                vectors = "fasttext.simple.300d"  # must handle str on Py2
```
Comment (zhangguanheng66) [inline test/test_vocab.py:102]: Remove the comment

File: torchtext/data/batch.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import torch
 
 
-class Batch(object):
+class Batch:
```
Comment (zhangguanheng66) [inline torchtext/data/batch.py:4]: Is this relevant to the drop of python2 support?

File: torchtext/data/dataset.py [hunk 8]
```diff
@@ -278,7 +276,7 @@ def __init__(self, path, format, fields, skip_header=False,
                 else:
                     fields.append(field)
 
-        super(TabularDataset, self).__init__(examples, fields, **kwargs)
```
Comment (zhangguanheng66) [inline torchtext/data/dataset.py:281]: Is this relevant to the drop of python2 support?

File: torchtext/data/dataset.py [hunk 9]
```diff
@@ -315,7 +313,7 @@ def stratify(examples, strata_field):
     # The field has to be hashable otherwise this doesn't work
     # There's two iterations over the whole dataset here, which can be
     # reduced to just one if a dedicated method for stratified splitting is used
-    unique_strata = set(getattr(example, strata_field) for example in examples)
```
Comment (zhangguanheng66) [inline torchtext/data/dataset.py:318]: Is this relevant to the drop of python2 support?

File: torchtext/data/example.py [hunk 10]
```diff
@@ -1,9 +1,8 @@
-import six
 import json
 from functools import reduce
 
 
-class Example(object):
```
Comment (zhangguanheng66) [inline torchtext/data/example.py:6]: Is this relevant to the drop of python2 support?

File: torchtext/data/field.py [hunk 11]
```diff
@@ -11,7 +9,7 @@
 from ..vocab import Vocab, SubwordVocab
 
 
-class RawField(object):
```
Comment (zhangguanheng66) [inline torchtext/data/field.py:14]: Is this relevant to the drop of python2 support?

File: torchtext/data/field.py [hunk 12]
```diff
@@ -378,7 +372,7 @@ def __init__(self, **kwargs):
             kwargs['tokenize'] = 'revtok'
         if 'unk_token' not in kwargs:
             kwargs['unk_token'] = ' UNK '
-        super(ReversibleField, self).__init__(**kwargs)
+        super().__init__(**kwargs)
```
Comment (zhangguanheng66) [inline torchtext/data/field.py:375]: Is this relevant to the drop of python2 support?

File: torchtext/data/iterator.py [hunk 13]
```diff
@@ -286,8 +284,6 @@ def pool(data, batch_size, key, batch_size_fn=lambda new, count, sofar: count,
             if sort_within_batch \
             else batch(p, batch_size, batch_size_fn)
         if shuffle:
-            for b in random_shuffler(list(p_batch)):
```
Comment (zhangguanheng66) [inline torchtext/data/iterator.py:289]: Is this relevant to the drop of python2 support?

File: torchtext/datasets/babi.py [hunk 14]
```diff
@@ -59,11 +58,11 @@ def __init__(self, path, text_field, only_supporting=False, **kwargs):
         fields = [('story', text_field), ('query', text_field), ('answer', text_field)]
         self.sort_key = lambda x: len(x.query)
 
-        with open(path, 'r', encoding="utf-8") as f:
+        with open(path, encoding="utf-8") as f:
```
Comment (zhangguanheng66) [inline torchtext/datasets/babi.py:61]: Is this relevant to the drop of python2 support?

File: torchtext/experimental/datasets/text_classification.py [hunk 15]
```diff
@@ -37,7 +36,7 @@ def _generate_data_iterators(dataset_name, root, ngrams, tokenizer, data_select)
     if not tokenizer:
         tokenizer = get_tokenizer("basic_english")
 
-    if not set(data_select).issubset(set(('train', 'test'))):
+    if not set(data_select).issubset({'train', 'test'}):
```
Comment (zhangguanheng66) [inline torchtext/experimental/datasets/text_classification.py:39]: Is this relevant to the drop of python2 support?

File: examples/text_classification/iterable_train.py (hunk 1 above)
Comment (hugovk) [reply examples/text_classification/iterable_train.py:150]: In Python 3, `io.open` is an alias for builtin `open`:

https://docs.python.org/3/library/io.html#io.open

File: examples/text_classification/iterable_train.py (hunk 3 above)
Comment (hugovk) [reply examples/text_classification/iterable_train.py:212]: Yes, it's using Python 3 syntax that was not possible also when supporting Python 2.

File: torchtext/data/dataset.py (hunk 9 above)
Comment (hugovk) [reply torchtext/data/dataset.py:318]: Yes, it's using Python 3 syntax that was not possible when also supporting Python 2.

File: torchtext/data/iterator.py (hunk 13 above)
Comment (hugovk) [reply torchtext/data/iterator.py:289]: Yes, it's using Python 3 syntax that was not possible when also supporting Python 2.

File: torchtext/datasets/babi.py (hunk 14 above)
Comment (hugovk) [reply torchtext/datasets/babi.py:61]: Yes, it's using Python 3 syntax that was not possible when also supporting Python 2.

File: torchtext/experimental/datasets/text_classification.py (hunk 15 above)
Comment (hugovk) [reply torchtext/experimental/datasets/text_classification.py:39]: Yes, it's using Python 3 syntax that was not possible when also supporting Python 2.

File: examples/text_classification/iterable_train.py (hunk 2 above)
Comment (hugovk) [reply examples/text_classification/iterable_train.py]: Reverted. Please could you add a comment on which change you're referring to below?

File: torchtext/data/batch.py (hunk 7 above)
Comment (hugovk) [reply torchtext/data/batch.py]: Reverted.

File: torchtext/data/field.py (hunk 11 above)
Comment (zhangguanheng66) [inline torchtext/data/field.py:14]: Can you revert those similar changes throughout the PR?

File: torchtext/data/field.py [hunk 16]
```diff
@@ -378,7 +372,7 @@ def __init__(self, **kwargs):
             kwargs['tokenize'] = 'revtok'
         if 'unk_token' not in kwargs:
             kwargs['unk_token'] = ' UNK '
-        super(ReversibleField, self).__init__(**kwargs)
```
Comment (zhangguanheng66) [inline torchtext/data/field.py:381]: Can you revert those similar changes throughout the PR?

**Comment by hugovk** [comment]
I've reverted the changes made by [pyupgrade](https://github.com/asottile/pyupgrade/) and instead only removed six and `__future__`.

Please let me know if you'd like any of the other pyupgrade changes re-applied.

**Comment by zhangguanheng66** [comment]
> I've reverted the changes made by [pyupgrade](https://github.com/asottile/pyupgrade/) and instead only removed six and `__future__`.
> 
> Please let me know if you'd like any of the other pyupgrade changes re-applied.

Thanks for the contribution. LGTM.  For now, let's make those necessary changes only.

File: test/data/test_dataset.py [hunk 17]
```diff
@@ -246,7 +244,7 @@ def test_csv_dataset_quotechar(self):
 
         with tempfile.NamedTemporaryFile(dir=self.test_dir) as f:
             for example in example_data:
-                f.write(six.b("{}\n".format(",".join(example))))
+                f.write("{}\n".format(",".join(example)).encode("latin-1"))
```
Comment (cpuhrsch) [inline test/data/test_dataset.py:247]: Where is the need for encode coming from?

Comment (cpuhrsch) [reply test/data/test_dataset.py:247]: In particular latin-1 vs. UTF-8?

Comment (hugovk) [reply test/data/test_dataset.py:247]: From the six documentation:

> `six.b`_(data)_
>
> A “fake” bytes literal. _data_ should always be a normal string literal. ... In Python 3, _data is_ encoded with the latin-1 encoding to bytes.

https://six.readthedocs.io/#six.b

From the six code:
```python
if PY3:
    def b(s):
        return s.encode("latin-1")
```
